### PR TITLE
replace config with protobuf

### PR DESCRIPTION
**Pull request by Shikugawa**
It also includes

- pretty refactor of circular buffer
- introduce segment context factory

### CODE REVIEW

File: cpp2sky/config.proto [hunk 1]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+message SegmentConfig {
```
Comment (wu-sheng) [inline cpp2sky/config.proto:3]: I feel these two belong to tracer.

Comment (Shikugawa) [reply cpp2sky/config.proto:3]: I'm worrying about to belong this on TracerConfig to be a confusion when integrating Envoy. (It won't use ClientConfig and TracerConfig, but SegmentConfig will be.)

Comment (wu-sheng) [reply cpp2sky/config.proto:3]: Oh, I see. Then good to me.

**Comment by Shikugawa** [comment]
@wu-sheng Changed to use only `TracerConfig`, is it look better?

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**